### PR TITLE
backend/refactor: #BKN-2938 issue-report

### PR DESCRIPTION
**Pull request by khuzema786**
## Type of Change

- [x] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description
- Seperation issue report list & issue report info API
- Updated DB Schema
- Refactoring
- Added validations

### Additional Changes

- [x] This PR modifies the database schema (database migration added)
- [x] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [x] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Dashboard/Issue.hs [hunk 1]
```diff
@@ -31,84 +47,152 @@ toCommonIssueStatus = \case
...
+            issueReportId,
+            authorId = cast justUserId,
+            comment = "Assignee Updated : " <> show justAssignee,
+            createdAt = now
           }
+
+issueAddComment :: ShortId DM.Merchant -> Id DIR.IssueReport -> Common.IssueAddCommentReq -> Flow APISuccess
+issueAddComment _merchantShortId issueReportId Common.IssueAddCommentReq {..} = do
+  case userId of
+    Just justUserId -> do
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Dashboard/Issue.hs:183]: userId <- justUserId & fromMaybe (InternalError "userId is not attatched to the request by dashboard.")

**Review by hkmangla — CHANGES_REQUESTED** [review]

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Dashboard/Message.hs [hunk 2]
```diff
@@ -101,9 +102,9 @@ uploadFile merchantShortId Common.UploadFileRequest {..} = do
   merchant <- findMerchantByShortId merchantShortId
   mediaFile <- L.runIO $ base64Encode <$> BS.readFile file
   filePath <- createFilePath merchant.id.getId fileType "" -- TODO: last param is extension (removed it as the content-type header was not comming with proxy api)
-  mediaFileUrlPattern <- asks (.mediaFileUrlPattern)
+  transporterConfig <- CQTC.findByMerchantId merchant.id >>= fromMaybeM (MerchantServiceUsageConfigNotFound merchant.id.getId)
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Dashboard/Message.hs:105]: add proper error type.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Issue.hs [hunk 3]
```diff
@@ -126,22 +131,51 @@ createMediaEntry url fileType = do
 
 issueMediaUpload :: Id SP.Person -> Common.IssueMediaUploadReq -> Flow Common.IssueMediaUploadRes
 issueMediaUpload driverId Common.IssueMediaUploadReq {..} = do
+  contentType <- validateContentType
+  fileSize <- L.runIO $ withFile file ReadMode hFileSize
+  person <- Esq.runInReplica $ QP.findById driverId >>= fromMaybeM (PersonNotFound driverId.getId)
+  transporterConfig <- CQTC.findByMerchantId person.merchantId >>= fromMaybeM (MerchantServiceUsageConfigNotFound person.merchantId.getId)
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Issue.hs:137]: proper error type.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Environment.hs [hunk 4]
```diff
@@ -142,11 +141,16 @@ data AppEnv = AppEnv
     driverLocationUpdateTopic :: Text,
     broadcastMessageTopic :: Text,
     snapToRoadSnippetThreshold :: HighPrecMeters,
-    mediaFileUrlPattern :: Text,
     minTripDistanceForReferralCfg :: Maybe HighPrecMeters
   }
   deriving (Generic)
 
+data MediaFileConfig = MediaFileConfig
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Environment.hs:148]: remove this.

**Review by hkmangla — DISMISSED** [review]

File: Backend/app/dashboard/CommonAPIs/src/Dashboard/Common/Issue.hs [hunk 5]
```diff
@@ -22,13 +22,34 @@ import Servant hiding (Summary)
 
 -- we need to save endpoint transactions only for POST, PUT, DELETE APIs
 data IssueEndpoint
-  = IssueListEndpoint
+  = IssueCategoryListEndpoint
+  | IssueListEndpoint
+  | IssueInfoEndpoint
```
Comment (roman-bodavskiy) [inline Backend/app/dashboard/CommonAPIs/src/Dashboard/Common/Issue.hs:27]: this datatype used for storing transaction in DB and not required for GET APIs. So it should be:

```
data IssueEndpoint
  = IssueUpdateEndpoint
  | IssueAddCommentEndpoint
```

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Issue.hs [hunk 6]
```diff
@@ -151,25 +185,76 @@ createIssueReport driverId Common.IssueReportReq {..} = do
...
+        Common.IssueInfoRes
+          { issueReportId = cast issueReport.id,
+            category = fromMaybe issueCategory.category $ (.translation) <$> issueCategoryTranslation,
+            option = Nothing,
+            assignee = issueReport.assignee,
+            description = issueReport.description,
+            status = toCommonIssueStatus issueReport.status,
+            mediaFiles = mkMediaFiles mediaFiles,
+            createdAt = issueReport.createdAt
+          }
```
Comment (roman-bodavskiy) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Issue.hs:233]: This code is duplicated. All fields, except of `option` are the same

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Tools/Error.hs [hunk 7]
```diff
@@ -246,3 +246,94 @@ instance IsBaseError FareParametersError where
...
+
+instanceExceptionWithParent 'HTTPException ''IssueReportError
+
+instance IsBaseError IssueReportError where
+  toMessage = \case
+    IssueReportNotFound issueReportId -> Just $ "IssueReport with issueReportId \"" <> show issueReportId <> "\" not found."
+
+instance IsHTTPError IssueReportError where
+  toErrorCode (IssueReportNotFound _) = "ISSUE_REPORT_NOT_FOUND"
+  toHttpCode (IssueReportNotFound _) = E400
```
Comment (roman-bodavskiy) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Tools/Error.hs:262]: this should be E500

File: Backend/dev/migrations/dynamic-offer-driver-app/0128-change-issue-table-properties.sql [hunk 8]
```diff
@@ -0,0 +1,40 @@
+-- comments table
+DELETE FROM atlas_driver_offer_bpp.comment;
+ALTER TABLE atlas_driver_offer_bpp.comment DROP COLUMN author;
+ALTER TABLE atlas_driver_offer_bpp.comment ADD COLUMN author_id character varying(255) NOT NULL;
```
Comment (roman-bodavskiy) [inline Backend/dev/migrations/dynamic-offer-driver-app/0128-change-issue-table-properties.sql:4]: `character(36)` here and in other references

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Issue.hs (hunk 6 above)
Comment (roman-bodavskiy) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Issue.hs:233]: I meant, you can avoid duplication like this:

```
  mbIssueOption <- (join <$>) $ forM issueReport.optionId $ \justIssueOption -> do
    CQIO.findByIdAndLanguage justIssueOption language
  pure $
    Common.IssueInfoRes
      { issueReportId = cast issueReport.id,
        category = fromMaybe issueCategory.category $ (.translation) <$> issueCategoryTranslation,
        option = mkIssueOption <$> mbIssueOption,
        assignee = issueReport.assignee,
        description = issueReport.description,
        status = toCommonIssueStatus issueReport.status,
        mediaFiles = mkMediaFiles mediaFiles,
        createdAt = issueReport.createdAt
      }
```

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Issue.hs [hunk 9]
```diff
@@ -25,70 +28,72 @@ import Kernel.Utils.Common
...
+toCommonIssueStatus = \case
+  D.NEW -> Common.NEW
+  D.INPROGRESS -> Common.INPROGRESS
+  D.RESOLVED -> Common.RESOLVED
+
+getLanguage :: Id SP.Person -> Maybe Language -> Flow Language
+getLanguage driverId mbLanguage = do
+  driverDetail <- Esq.runInReplica $ QP.findById driverId
+  return $
+    fromMaybe ENGLISH $
```
Comment (piyushKumar-1) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Issue.hs:48]: return $ fromMaybe ENGLISH ((. language) =<< driverDetail)

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Issue.hs [hunk 10]
```diff
@@ -151,25 +185,64 @@ createIssueReport driverId Common.IssueReportReq {..} = do
...
+        mediaFiles = mkMediaFiles mediaFiles,
+        createdAt = issueReport.createdAt
+      }
+  where
+    mkIssueOption :: (D.IssueOption, Maybe D.IssueTranslation) -> Text
+    mkIssueOption (issueOption, issueOptionTranslation) =
+      fromMaybe issueOption.option $ (.translation) <$> issueOptionTranslation
+    mkMediaFiles :: [D.MediaFile] -> [Common.MediaFile_]
+    mkMediaFiles =
+      foldr'
```
Comment (piyushKumar-1) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Issue.hs:228]: can we move this function to the common lib.

**Review by piyushKumar-1 — APPROVED** [review]

**Pull request merged**